### PR TITLE
Implement Pager for auto-paging requests

### PR DESCRIPTION
**Pull request by t8y8**
Here's a first draft based on @RussTheAerialist's code in `pagination_sample.py`.

Currently it:
- Takes an endpoint as its argument instead of a `get` function, this is because these calls only apply to `get` anyway, so let's keep it clean.
- Takes an options object and seeds the filters/sorts with it, as well as the initial page size
- Can be used as an iterator

I would like it to:

- Not do everything in the `__call__` method. I _think_ we can move that to a helper or to `__init__`
    - We would make the first page call on instantiation, which populates some properties that could be useful (like `__len__`)
- Since we know the `total_avaliable` we might be able to implement `__len__`. This gives users the ability to check the length to see if they really want to iterate over it or not.
- I need to read more to see if I need to worry about `__next__` or not.
- Add tests

This script works (tested manually):
```python
import tableauserverclient as TSC

auth_stuff = TSC.TableauAuth(username='USERNAME', password='PASSWORD')

server = TSC.Server("http://SERVER")
server.auth.sign_in(auth_stuff)

options = TSC.RequestOptions()
options.filter.add(TSC.Filter(TSC.RequestOptions.Field.Name,
                                         TSC.RequestOptions.Operator.Equals,
                                         "WORKBOOKNAME"))

# Call with a filter
wbs = TSC.server.Pager(server.workbooks, options)
print(len([i for i in wbs[)) # prints 1

# Get them all!
wbs = TSC.server.Pager(server.workbooks)
print(len([i for i in wbs])) # prints 1086 on my test server, that's all of em!
server.auth.sign_out()
```

Open Questions:

1. Where should it live? I'm not sure Server is the right place
2. Thoughts on implementing `__len__`?
3. Thoughts on implementing the primary logic somewhere other than `__call__`? It might just be me who thinks it's weird.

/cc @RussTheAerialist @LGraber

### CODE REVIEW

File: tableauserverclient/server/server.py [hunk 1]
```diff
@@ -1,8 +1,47 @@
...
+
+    server is the server object that will be used when calling the callback.  It will be passed
+    to the callback on each iteration
+
+    Callback is expected to take a server object and a request options and return two values, an array of results,
+    and the pagination item from the current call.  This will be used to build subsequent requests.
+    """
+
+    def __init__(self, fetcher, opts=None):
+        self._fetcher = fetcher.get
```
Comment (RussTheAerialist) [inline tableauserverclient/server/server.py]: probably name the argument that is passed in `endpoint`

File: tableauserverclient/server/server.py [hunk 2]
```diff
@@ -1,8 +1,47 @@
...
+        self._options = opts
+
+    def __call__(self):
```
Comment (RussTheAerialist) [inline tableauserverclient/server/server.py]: yeah this is not ideal.  It should probably just live in `__iter__` and we get rid of `__call__`

File: tableauserverclient/server/server.py [hunk 3]
```diff
@@ -1,8 +1,47 @@
...
+        current_item_list, last_pagination_item = self._fetcher(self._options)
```
Comment (RussTheAerialist) [inline tableauserverclient/server/server.py]: this won't work if they set a page number in the options because count will never be greater than total available until we hit the end and throw an error.

We should check that a page number hasn't been set in the options, and if it has output a warning or throw an exception.

Comment (t8y8) [inline tableauserverclient/server/server.py]: There's always a page number (it's defaulted to 1).

Should the logic be:

1. On first iteration, run the generator
2. Store some state that says "we're working on it now"
3. Then just exhaust the generator starting from that point, then they could, if they want, start on page 3

File: tableauserverclient/server/server.py [hunk 4]
```diff
@@ -1,8 +1,47 @@
...
 from .endpoint import Sites, Views, Users, Groups, Workbooks, Datasources, Projects, Auth, Schedules, ServerInfo
+from . import RequestOptions
+
 import requests
 
 
+class Pager(object):
+    """ This class returns a generator that will iterate over all of the results.
+
+    server is the server object that will be used when calling the callback.  It will be passed
```
Comment (LGraber) [inline tableauserverclient/server/server.py]: This comment does not appear to be relevant anymore

Comment (LGraber) [inline tableauserverclient/server/server.py]: why doesn't it work if we simply change how we initialize count?

File: tableauserverclient/server/server.py [hunk 5]
```diff
@@ -1,8 +1,47 @@
...
+
+    def __call__(self):
+        current_item_list, last_pagination_item = self._fetcher(self._options)
+        count = 0
+
+        while count < last_pagination_item.total_available:
+            if len(current_item_list) == 0:
+                current_item_list, last_pagination_item = self._load_next_page(current_item_list, last_pagination_item)
+
+            yield current_item_list.pop(0)
```
Comment (LGraber) [inline tableauserverclient/server/server.py]: This needs a bit more protection. Technically _load_next_page can return an empty list ... even after we fix our counting issue. This could happen if someone deletes an item while we are iterating. We should make sure the list is not empty before popping.

**Comment by LGraber** [comment]
Oh yeah ... why did you put this in the server.py file and not its own file?

**Comment by t8y8** [comment]
Alright! Addressed feedback (in person, and on here) tests added, and tested manually with this code:
```python
import tableauserverclient as TSC


auth_stuff = TSC.TableauAuth(username='USER', password='password')

server = TSC.Server("http://server")

server.auth.sign_in(auth_stuff)

print(server.auth_token)

options = TSC.RequestOptions(pagenumber=2, pagesize=300)

wbs = TSC.Pager(server.workbooks, options)
print("GOT WBS")
print(len({i.id for i in wbs}))
server.auth.sign_out()
```

File: test/test_pager.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            m.get(self.baseurl + "?pageNumber=1&pageSize=1", text=page_1)
+            m.get(self.baseurl + "?pageNumber=2&pageSize=1", text=page_2)
+            m.get(self.baseurl + "?pageNumber=3&pageSize=1", text=page_3)
+
+            # No options should get all 3
+            workbooks = TSC.Pager(self.server.workbooks)
+            self.assertTrue(len(list(workbooks)) == 3)
+
+            # Let's check that workbook items aren't duplicates
+            workbooks = TSC.Pager(self.server.workbooks)
```
Comment (LGraber) [inline test/test_pager.py]: why did you need to call this again?

File: test/test_pager.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        with open(GET_XML_PAGE3, 'rb') as f:
+            page_3 = f.read().decode('utf-8')
+        with requests_mock.mock() as m:
+            # Register Pager with default request options
+            m.get(self.baseurl, text=page_1)
+
+            # Register Pager with some pages
+            m.get(self.baseurl + "?pageNumber=1&pageSize=1", text=page_1)
+            m.get(self.baseurl + "?pageNumber=2&pageSize=1", text=page_2)
+            m.get(self.baseurl + "?pageNumber=3&pageSize=1", text=page_3)
```
Comment (LGraber) [inline test/test_pager.py]: why are you registering these if we are not supposed to call them?

File: test/test_pager.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    def test_pager_with_options(self):
+        with open(GET_XML_PAGE1, 'rb') as f:
+            page_1 = f.read().decode('utf-8')
+        with open(GET_XML_PAGE2, 'rb') as f:
+            page_2 = f.read().decode('utf-8')
+        with open(GET_XML_PAGE3, 'rb') as f:
+            page_3 = f.read().decode('utf-8')
+        with requests_mock.mock() as m:
+            # Register Pager with default request options
+            m.get(self.baseurl, text=page_1)
```
Comment (LGraber) [inline test/test_pager.py]: Why are you registering this if the test is not supposed to call it?

**Comment by LGraber** [comment]
I am confused because the tests don't represent how we expect the users to use this class (ie you keep casting to list and doing len) and you don't have a sample of how to use it. Can you please add a sample

File: test/test_pager.py (hunk 6 above)
Comment (RussTheAerialist) [inline test/test_pager.py]: generators are one time use.  Once you iterate over them, you cannot go back.  He could have stored the output of `list(workbooks)` and then he wouldn't have had to call it twice.

File: test/test_pager.py (hunk 7 above)
Comment (LGraber) [inline test/test_pager.py]: Hahahaha. :) I know what it does. My comment was that we don't call these urls in this test so why do we need to register them. :)

File: test/test_pager.py (hunk 6 above)
Comment (RussTheAerialist) [inline test/test_pager.py]: though, it might be better (and more instructive) to use

```python
self.assertEquals(3, itertools.count(workbooks)
```

File: test/test_pager.py (hunk 7 above)
Comment (LGraber) [inline test/test_pager.py]: My bad ... it does get called (except maybe the first one). OOps

File: test/test_pager.py [hunk 9]
```diff
@@ -37,11 +37,10 @@ def test_pager_with_no_options(self):
             m.get(self.baseurl + "?pageNumber=3&pageSize=1", text=page_3)
 
             # No options should get all 3
-            workbooks = TSC.Pager(self.server.workbooks)
-            self.assertTrue(len(list(workbooks)) == 3)
+            workbooks = list(TSC.Pager(self.server.workbooks))
+            self.assertTrue(len(workbooks) == 3)
```
Comment (RussTheAerialist) [inline test/test_pager.py]: nitpick
```python
self.assertEquals(len(workbooks), 3)
```
